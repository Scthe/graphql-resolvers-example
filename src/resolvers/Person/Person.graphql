extend type Query {
  person(id: ID!): Person!
  people(name: String): PeoplesList!
}

enum Gender {
  MALE
  FEMALE
}

type Person {
  id: ID!
  name: String!
  castCredits: ShowCharactersList!
  countryOfBirth: String
  birthday: String # TODO date
  deathday: String # TODO date
  gender: Gender
}

type PeoplesList {
  node: [Person!]!
  meta: ListMeta!
}
