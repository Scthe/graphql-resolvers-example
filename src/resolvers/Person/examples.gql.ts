export const GET_SINGLE_GQL = `{
  person(id:1){
    id
    name
    birthYear
    gender
  }
}`;