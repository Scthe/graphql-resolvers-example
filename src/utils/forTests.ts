import nock from "nock";

// utils used in tests

export const setupNock = (): void => {
  beforeEach(() => {
    // each jest test loads modules anew
    nock.disableNetConnect();
    nock.enableNetConnect("127.0.0.1");
  });

  afterEach(() => {
    nock.cleanAll();
  });

  afterAll(() => {
    nock.restore(); // https://github.com/nock/nock/issues/1817
  });
};
