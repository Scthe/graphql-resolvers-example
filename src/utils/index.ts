// eslint-disable @typescript-eslint/no-explicit-any
export const parseId = (id: number | string): ID => {
  if (typeof id === "string") {
    return parseInt(id, 10) as any;
  }
  return id as any;
};
