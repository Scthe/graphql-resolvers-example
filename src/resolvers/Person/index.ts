import { IResolvers } from "apollo-server";

import PersonQuery from "./queries/Person";
import PeopleQuery from "./queries/People";
import Person from "./types/Person";
import PeoplesList from "./types/PeoplesList";

export default {
  Query: {
    person: PersonQuery,
    people: PeopleQuery,
  },
  Person,
  PeoplesList,
} as IResolvers<any, any>; // eslint-disable-line @typescript-eslint/no-explicit-any
