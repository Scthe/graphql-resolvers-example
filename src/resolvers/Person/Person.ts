import GqlContext from "GqlContext";
import { copyFromRestResponse } from "utils";

type RootType = ID;

const getItem = async (id: RootType, context: GqlContext) => {
  return context.dataSources.personAPI.getPerson(id);
};

async function gender(
  root: RootType,
  _: any,
  context: GqlContext
): Promise<string> {
  const item = await getItem(root, context);
  switch (item.gender.toLowerCase()) {
    case "male":
      return "MALE";
    case "female":
      return "FEMALE";
    default:
      return "OTHER";
  }
}

export default {
  id: (root: RootType): ID => root,
  name: copyFromRestResponse(getItem, "name"),
  birthYear: copyFromRestResponse(getItem, "birth_year"),
  gender,
};
