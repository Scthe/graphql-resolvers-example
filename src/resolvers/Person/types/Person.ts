import GqlContext from "GqlContext";
import { Gender, PersonResolvers } from "typingsGql";
import { copyFromRestResponse } from "utils/graphql";

export type RootType = ID;

type ResolverType = PersonResolvers;

const getItem = (root: RootType, context: GqlContext) => {
  return context.dataSources.peopleAPI.getOne(root);
};

const countryOfBirth = async (
  root: RootType,
  _args: unknown,
  context: GqlContext
): Promise<string | null> => {
  const item = await getItem(root, context);
  return item.country?.name || null;
};

const gender = async (
  root: RootType,
  _args: unknown,
  context: GqlContext
): Promise<Gender | null> => {
  const item = await getItem(root, context);
  const gender = item.gender;
  if (gender === "Male") {
    return Gender.Male;
  } else if (gender === "Female") {
    return Gender.Female;
  }
  return null;
};

const resolver: ResolverType = {
  id: async (root: RootType) => root,
  name: copyFromRestResponse(getItem, "name"),
  // Bug in `graphql-codegen`: `birthday` is nullable in schema, but the generated code allows only `null` and not `undefined`
  birthday: copyFromRestResponse(getItem, "birthday") as any,
  deathday: copyFromRestResponse(getItem, "deathday") as any,
  gender,
  countryOfBirth,
  castCredits: async (root: RootType) => ({ personId: root }),
};

export default resolver;
