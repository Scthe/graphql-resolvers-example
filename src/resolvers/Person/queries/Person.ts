import { QueryPersonArgs } from "typingsGql";
import { RootType as PersonType } from "../types/Person";
import { parseId } from "utils";

export default (_root: any, args: QueryPersonArgs): PersonType =>
  parseId(args.id);
