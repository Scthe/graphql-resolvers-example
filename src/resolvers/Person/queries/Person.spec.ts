import supertest from "supertest";
import nock, { Scope } from "nock";
import merge from "deepmerge";

import app from "../../../index";
import * as gqlExamples from "./Person.example";
import { Person as GqlSchemaPerson } from "typingsGql";
import { setupNock } from "utils/forTests";

import PersonMock from "dataSources/tvmaze/mocks/person.mock.json";
import PersonCastCreditMock from "dataSources/tvmaze/mocks/personCastcredit.mock.json";
import ShowMock from "dataSources/tvmaze/mocks/show.mock.json";

describe("GraphQL: person()", () => {
  let scope: Scope | null = null;

  setupNock();

  afterEach(() => {
    scope && scope.done(); // verify all planned requests were actually done
  });

  const mockSuccessPersonResponse = (item: Partial<typeof PersonMock> = {}) => {
    // NOTE: this allows for ONE request
    scope = nock("https://api.tvmaze.com")
      .get(`/people/${item.id}`)
      .reply(200, {
        ...PersonMock,
        ...item,
      });
  };

  const mockSuccessChildrenResponse = (personId: number) => {
    const personCastCreditMock = (
      showId: number,
      characterId: number,
      characterName: string
    ) =>
      merge(PersonCastCreditMock, {
        _links: {
          show: {
            href: `https://api.tvmaze.com/shows/${showId}`,
          },
        },
        _embedded: {
          character: {
            id: characterId,
            name: characterName,
          },
        },
      });
    const showMock = (id: number, name: string) => ({
      ...ShowMock,
      id,
      name,
    });

    scope!
      .get(`/people/${personId}/castcredits?embed=character`)
      .reply(200, [
        personCastCreditMock(100, 1, "Dr. Gregory House"),
        personCastCreditMock(200, 5, "The Prince Regent"),
      ])
      .get(`/shows/100`)
      .reply(200, showMock(100, "House"))
      .get(`/shows/200`)
      .reply(200, showMock(200, "Blackadder"));
  };

  test("should only fetch what is needed", () => {
    // In this test we check if we overfetch. We query only for id (which we pass in args),
    // so no network request is expected (nock would throw).
    return supertest(app)
      .post("/graphql")
      .send({
        query: gqlExamples.GQL_PERSON_ONLY_ID,
      })
      .then((resp) => {
        const item = resp.body.data.person as GqlSchemaPerson;
        expect(item.id).toBe("" + PersonMock.id);
        expect(item).toMatchSnapshot(); // just for a good measure
      });
  });

  test("should fetch basic item properties", () => {
    // In this test we get all properties of a person (that are not a paginated list).
    // We should be able to deduplicate request and do it only once.
    mockSuccessPersonResponse({ id: 25439 });

    return supertest(app)
      .post("/graphql")
      .send({
        query: gqlExamples.GQL_PERSON_FIELDS,
      })
      .then((resp) => {
        const item = resp.body.data.person as GqlSchemaPerson;
        expect(item.id).toBe("25439");
        expect(item).toMatchSnapshot(); // just for a good measure
      })
      .finally();
  });

  test("should fetch show with all children depdendencies", () => {
    // This is quite extreme test. We will fetch person with all casting data.
    // In production app this test would be VERY brittle, so be careful.
    // What we want is to invoke as much resolvers as we can (person/cast/show).
    // TBH I'm doing this to showcase that all connections in graph work correctly.
    const personId = 25439;
    mockSuccessPersonResponse({ id: personId });
    mockSuccessChildrenResponse(personId);

    return supertest(app)
      .post("/graphql")
      .send({
        query: gqlExamples.GQL_PERSON_WITH_CHILDREN,
      })
      .then((resp) => {
        const item = resp.body.data.person as GqlSchemaPerson;
        expect(item.id).toBe("" + personId);
        expect(item.name).toBe(PersonMock.name);
        expect(item.castCredits.meta.totalCount).toBe(2);
        expect(item.castCredits.node).toHaveLength(2);
        // check cast character 1
        expect(item.castCredits.node[0].id).toBe("1");
        expect(item.castCredits.node[0].name).toBe("Dr. Gregory House");
        expect(item.castCredits.node[0].show.name).toBe("House");
        // check cast character 2
        expect(item.castCredits.node[1].id).toBe("5");
        expect(item.castCredits.node[1].name).toBe("The Prince Regent");
        expect(item.castCredits.node[1].show.name).toBe("Blackadder");

        // just for a good measure
        expect(item).toMatchSnapshot();
      })
      .finally();
  });
});
