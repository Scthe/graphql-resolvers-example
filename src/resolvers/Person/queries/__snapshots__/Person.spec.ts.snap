// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQL: person() should fetch basic item properties 1`] = `
Object {
  "birthday": "1959-06-11",
  "countryOfBirth": "United Kingdom",
  "deathday": null,
  "gender": "MALE",
  "id": "25439",
  "name": "Hugh Laurie",
}
`;

exports[`GraphQL: person() should fetch show with all children depdendencies 1`] = `
Object {
  "castCredits": Object {
    "meta": Object {
      "totalCount": 2,
    },
    "node": Array [
      Object {
        "id": "1",
        "name": "Dr. Gregory House",
        "person": Object {
          "id": "25439",
          "name": "Hugh Laurie",
        },
        "show": Object {
          "id": "100",
          "name": "House",
        },
      },
      Object {
        "id": "5",
        "name": "The Prince Regent",
        "person": Object {
          "id": "25439",
          "name": "Hugh Laurie",
        },
        "show": Object {
          "id": "200",
          "name": "Blackadder",
        },
      },
    ],
  },
  "id": "25439",
  "name": "Hugh Laurie",
}
`;

exports[`GraphQL: person() should only fetch what is needed 1`] = `
Object {
  "id": "25439",
}
`;
